src/ledger/LedgerEntryScope.h

```diff
@@ -0,0 +1,306 @@
+    ScopedLedgerEntry<S>& operator=(ScopedLedgerEntry<S> const& other);
+    ScopedLedgerEntry<S>& operator=(ScopedLedgerEntry<S>&& other);
+
+    LedgerKey key() const;
+
+    LedgerEntry const& read_in_scope(LedgerEntryScope<S> const& scope) const;
+    LedgerEntry& modify_in_scope(LedgerEntryScope<S> const& scope);
+
+    bool operator==(ScopedLedgerEntry const& other) const;
+    bool operator<(ScopedLedgerEntry const& other) const;
+};
+
+template <StaticLedgerEntryScope S> class ScopedOptionalLedgerEntry
+{
+    static constexpr StaticLedgerEntryScope staticScope = S;
+    using ScopeIdT = LedgerEntryScopeID<S>;
+    using ScopeT = LedgerEntryScope<S>;
+    std::optional<LedgerEntry> mEntry;
+
+    friend class LedgerEntryScope<S>;
+#define FRIEND_MACRO(OUTER_SCOPE, INNER_SCOPE) \
+    friend class LedgerEntryScope<StaticLedgerEntryScope::INNER_SCOPE>;
+    FOREACH_STATIC_LEDGER_ENTRY_SCOPE_INNER(S, FRIEND_MACRO)
+#undef FRIEND_MACRO
+
+    ScopedOptionalLedgerEntry(ScopeIdT scopeID,
+                              std::optional<LedgerEntry> const& entry);
+    ScopedOptionalLedgerEntry(ScopeIdT scopeID,
+                              std::optional<LedgerEntry>&& entry);
+
+  public:
+    LedgerEntryScopeID<S> const mScopeID;
+
+    // Must construct with a scope.
+    ScopedOptionalLedgerEntry<S>() = delete;
+
+    ScopedOptionalLedgerEntry(ScopedOptionalLedgerEntry<S> const& other);
+    ScopedOptionalLedgerEntry(ScopedOptionalLedgerEntry<S>&& other);
+    ScopedOptionalLedgerEntry<S>&
+    operator=(ScopedOptionalLedgerEntry<S> const& other);
+    ScopedOptionalLedgerEntry<S>&
+    operator=(ScopedOptionalLedgerEntry<S>&& other);
+
+    ScopedOptionalLedgerEntry(ScopedLedgerEntry<S> const& other);
+    ScopedOptionalLedgerEntry(ScopedLedgerEntry<S>&& other);
+
+    std::optional<LedgerKey> key() const;
+
+    std::optional<LedgerEntry> const&
+    read_in_scope(LedgerEntryScope<S> const& scope) const;
+    std::optional<LedgerEntry>&
+    modify_in_scope(LedgerEntryScope<S> const& scope);
+
+    bool operator==(ScopedOptionalLedgerEntry const& other) const;
+    bool operator<(ScopedOptionalLedgerEntry const& other) const;
+};
+
+using GlobalLedgerEntry =
+    ScopedLedgerEntry<StaticLedgerEntryScope::GLOBAL_PAR_APPLY_STATE>;
+using ThreadLedgerEntry =
+    ScopedLedgerEntry<StaticLedgerEntryScope::THREAD_PAR_APPLY_STATE>;
+using TxLedgerEntry =
+    ScopedLedgerEntry<StaticLedgerEntryScope::TX_PAR_APPLY_STATE>;
+
+using GlobalOptionalLedgerEntry =
+    ScopedOptionalLedgerEntry<StaticLedgerEntryScope::GLOBAL_PAR_APPLY_STATE>;
+using ThreadOptionalLedgerEntry =
+    ScopedOptionalLedgerEntry<StaticLedgerEntryScope::THREAD_PAR_APPLY_STATE>;
+using TxOptionalLedgerEntry =
+    ScopedOptionalLedgerEntry<StaticLedgerEntryScope::TX_PAR_APPLY_STATE>;
+
+// Defines valid scope transitions as (DEST, SOURCE) pairs for compile-time
+// enforcement.
+#define FOR_EACH_VALID_SCOPE_ADOPTION(MACRO) \
```

Comment:
I think everything else in this file is pretty much template boilerplate, but this is very important to anyone dealing with ledger state and transitions. Can we add more comments/instructions here on how to modify this list, and maybe make it a little more prominent in the file? This might be worth it's own .h file just so we have a very clear representation of the data dependency graph.